pyro/poutine/posterior_messenger.py

```diff
@@ -0,0 +1,12 @@
+from .replay_messenger import ReplayMessenger
+
+
+class SamplePosteriorMessenger(ReplayMessenger):
```

Comment:
Now that I am thinking about it, I am not sure if this would work for the second case that you wanted to handle, since in that case what we need to do is to run the guide forward using the test data to get the correct batch shapes and then align them with the model. Replaying `cond_indep_stack` wouldn't solve this issue, as far as I can tell. Maybe we should write a test for this?

Reply:
Thanks for bringing it up.
I actually also discovered the issue yesterday 😄 (but I was doing some work-related programming first). I will see how to fix the issue.

Reply:
No worries! Also, there is no need to address everything in this PR. `AbstractInfer` is due a bigger refactor, so this will help clarify what should go into that refactor (in addition to #1725 for instance). I think that making those classes lazy might be the solution to this issue.

Reply:
Actually, I figured out how to do the fix before you responded, so I have tested it and pushed 😄. But a large refactoring to make it work better in the future would be very nice indeed!